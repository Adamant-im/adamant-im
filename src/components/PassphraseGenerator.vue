<template>
  <div :class="className">
    <div class="text-xs-center">
      <h3 :class="`${className}__title`">{{ $t('login.create_address_label') }}</h3>
      <v-btn @click="generatePassphrase" :class="`${className}__button`" flat small>
        {{ $t('login.new_button') }}
      </v-btn>
    </div>

    <transition name="slide-fade">
      <div :class="`${className}__box`" v-if="showPassphrase">
        <div
          v-html="$t('login.new_passphrase_label')"
          class="caption grey--text mt-2"
          ref="el"
        ></div>
        <v-textarea
          :value="passphrase"
          @click.prevent="selectText"
          type="text"
          multi-line
          readonly
          rows="3"
          class="pt-0"
          color="grey"
          no-resize
          ref="textarea"
        >
          <template slot="append">
            <div :class="`${className}__icons`">
              <icon
                @click="copyToClipboard"
                :width="24"
                :height="24"
                shape-rendering="crispEdges"
                :title="$t('login.copy_button_tooltip')"
              >
                <copy-icon/>
              </icon>
              <icon
                @click="saveFile"
                :width="24"
                :height="24"
                shape-rendering="auto"
                :title="$t('login.save_button_tooltip')"
              >
                <save-icon/>
              </icon>
              <icon
                @click="showQrcodeRendererDialog = true"
                :width="24"
                :height="24"
                shape-rendering="crispEdges"
                :title="$t('login.save_qr_code_tooltip')"
              >
                <qr-code-icon/>
              </icon>
            </div>
          </template>
        </v-textarea>
      </div>
    </transition>

    <QrcodeRendererDialog
      v-model="showQrcodeRendererDialog"
      :passphrase="passphrase"
    />
  </div>
</template>

<script>
import * as bip39 from 'bip39'

import { copyToClipboard, downloadFile } from '@/lib/textHelpers'
import QrcodeRendererDialog from '@/components/QrcodeRendererDialog'
import Icon from '@/components/icons/BaseIcon'
import CopyIcon from '@/components/icons/common/Copy'
import SaveIcon from '@/components/icons/common/Save'
import QrCodeIcon from '@/components/icons/common/QrCode'

export default {
  computed: {
    className () {
      return 'passphrase-generator'
    }
  },
  data: () => ({
    passphrase: '',
    showPassphrase: false,
    showQrcodeRendererDialog: false
  }),
  methods: {
    copyToClipboard () {
      copyToClipboard(this.passphrase)

      this.$emit('copy')
    },
    saveFile () {
      var iOS = /iPad|iPhone|iPod/.test(navigator.userAgent) && !window.MSStream
      if (!iOS) {
        downloadFile(
          this.passphrase,
          'adm-' + btoa(new Date().getTime()).replace('==', '') + '.txt',
          'text/plain'
        )
      }

      this.$emit('save')
    },
    selectText () {
      this.$refs.textarea.$el
        .querySelector('textarea')
        .select()
    },
    generatePassphrase () {
      this.passphrase = bip39.generateMnemonic()

      this.showPassphrase = true

      // callback after Vue rerender
      setTimeout(() => {
        this.$scrollTo(this.$refs.textarea.$el, { easing: 'ease-in' })
      }, 0)
    }
  },
  components: {
    Icon,
    CopyIcon,
    SaveIcon,
    QrCodeIcon,
    QrcodeRendererDialog
  }
}
</script>

<style lang="stylus" scoped>
@import '~vuetify/src/stylus/settings/_variables.styl'
@import '../assets/stylus/settings/_colors.styl'

/**
 * 1. Blur icons when focus textarea.
 * 2. Remove textarea border bottom.
 */
.passphrase-generator
  &__title
    font-size: 14px
    font-weight: 300
  &__button
    font-size: 16px
    font-weight: 500
    text-decoration: underline
    text-transform: unset
  &__box
    margin-top: 30px
    >>> .v-textarea textarea
      font-size: 14px
      font-weight: 300
    >>> .v-textarea
      .v-input__slot:before, .v-input__slot:after
        border-width: 0 // [2]

  &__icons
    margin-top: 10px

    > *:not(:first-child)
      margin-left: 8px

  .v-input--is-focused
    .v-icon
      opacity: 0.3 // [1]

/** Themes **/
.theme--light
  .passphrase-generator
    &__title, &__button
      color: inherit
    >>> .v-textarea textarea
      color: $adm-colors.regular
.theme--dark
  .passphrase-generator
    &__title, &__button
      color: $shades.white
</style>
