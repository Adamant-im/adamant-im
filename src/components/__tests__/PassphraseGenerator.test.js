import { shallowMount } from '@vue/test-utils'
import Vue from 'vue'
import VueI18n from 'vue-i18n'
import Vuetify from 'vuetify'
import PassphraseGenerator from '@/components/PassphraseGenerator'

import mockupI18n from './__mocks__/plugins/i18n'

Vue.use(VueI18n)
Vue.use(Vuetify)

describe('PassphraseGenerator.vue', () => {
  let i18n = null

  beforeEach(() => {
    i18n = mockupI18n()
  })

  it('renders the correct markup', () => {
    const wrapper = shallowMount(PassphraseGenerator, {
      i18n
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('check default data', () => {
    const wrapper = shallowMount(PassphraseGenerator, {
      i18n
    })

    expect(wrapper.vm.passphrase).toBe('')
    expect(wrapper.vm.showPassphrase).toBe(false)
  })

  it('should emit "copy" when copyToClipboard()', () => {
    // Mockup document.execCommand
    Object.defineProperty(document, 'execCommand', {
      value: jest.fn()
    })

    const wrapper = shallowMount(PassphraseGenerator, {
      i18n
    })

    wrapper.vm.passphrase = 'passphrase'
    wrapper.vm.copyToClipboard()

    expect(wrapper.emitted().copy).toBeTruthy()
    expect(document.execCommand).toBeCalled()
  })

  it('should emit "save" when saveFile()', () => {
    // Mockup window.URL
    Object.defineProperty(window, 'URL', {
      value: {
        createObjectURL: jest.fn()
      }
    })

    const wrapper = shallowMount(PassphraseGenerator, {
      i18n
    })

    wrapper.vm.passphrase = 'passphrase'
    wrapper.vm.saveFile()

    expect(wrapper.emitted().save).toBeTruthy()
    expect(window.URL.createObjectURL).toBeCalled()
  })

  it('should generate passphrase when generatePassphrase()', () => {
    const wrapper = shallowMount(PassphraseGenerator, {
      i18n
    })

    wrapper.vm.generatePassphrase()

    expect(wrapper.vm.showPassphrase).toBe(true)
    expect(wrapper.vm.passphrase.length > 0).toBe(true)
  })

  // @todo trigger('click') tests
})
