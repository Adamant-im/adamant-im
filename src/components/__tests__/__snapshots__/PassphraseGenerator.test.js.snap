// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PassphraseGenerator.vue > renders the correct markup 1`] = `
<div
  class="passphrase-generator"
  data-v-f3fe94d6=""
>
  <div
    class="text-center"
    data-v-f3fe94d6=""
  >
    <h3
      class="a-text-regular"
      data-v-f3fe94d6=""
    >
      Or create a new anonymous address:
    </h3>
    <v-btn
      class="a-btn-link mt-2"
      data-v-f3fe94d6=""
      size="small"
      variant="text"
    >
      Create new
    </v-btn>
  </div>
  <transition-stub
    appear="false"
    css="true"
    data-v-f3fe94d6=""
    name="slide-fade"
    persisted="false"
  >
    <!--v-if-->
  </transition-stub>
  <qrcode-renderer-dialog-stub
    data-v-f3fe94d6=""
    modelvalue="false"
    text=""
  />
</div>
`;
